lib/graphql_metrics/analyzer.rb

```diff
@@ -0,0 +1,140 @@
+      query_extracted(@static_query_metrics.merge(runtime_query_metrics))
+    end
+
+    def on_enter_operation_definition(_node, _parent, visitor)
+      @static_query_metrics = {
+        operation_type: visitor.query.selected_operation.operation_type,
+        operation_name: visitor.query.selected_operation.name,
+      }
+    end
+
+    def on_leave_field(node, _parent, visitor)
+      # NOTE: @rmosolgo "I think it could be reduced to `arguments = visitor.arguments_for(ast_node)`"
+      arguments = visitor.arguments_for(node, visitor.field_definition)
+      extract_arguments(arguments.argument_values.values, visitor.field_definition)
+
+      static_metrics = {
+        field_name: node.name,
+        return_type_name: visitor.type_definition.name,
+        parent_type_name: visitor.parent_type_definition.name,
+        deprecated: visitor.field_definition.deprecation_reason.present?,
+        path: visitor.response_path,
+      }
+
+      if GraphQLMetrics.timings_capture_enabled?(visitor.query.context)
+        @static_field_metrics << static_metrics
+      else
+        field_extracted(static_metrics)
+      end
+    end
+
+    def extract_fields_with_runtime_metrics(context)
+      @static_field_metrics.each do |static_metrics|
+        context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+          resolver_timings = ns[GraphQLMetrics::INLINE_FIELD_TIMINGS][static_metrics[:path]]
+          lazy_field_timing = ns[GraphQLMetrics::LAZY_FIELD_TIMINGS][static_metrics[:path]]
+
+          metrics = static_metrics.merge(
+            resolver_timings: resolver_timings,
+            lazy_resolver_timings: lazy_field_timing,
+          )
+
+          field_extracted(metrics)
+        end
+      end
+    end
+
+    def result
+      unless GraphQLMetrics.timings_capture_enabled?(@query.context)
+        # NOTE: If we're running as a static analyzer (i.e. not with instrumentation and tracing), we still need to
+        # flush static query metrics somewhere other than `after_query`.
+        @query.context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+          analyzer = ns[GraphQLMetrics::ANALYZER_INSTANCE_KEY]
+          analyzer.extract_query(context: @query.context)
+        end
+      end
+    end
+
+    private
+
+    def extract_arguments(value, field_defn)
+      case value
+      when Array
+        value.each do |v|
+          extract_arguments(v, field_defn)
+        end
+      when Hash
+        value.each_value do |v|
+          extract_arguments(v, field_defn)
+        end
+      when ::GraphQL::Query::Arguments
+        value.each_value do |arg_val|
+          extract_arguments(arg_val, field_defn)
+        end
+      when ::GraphQL::Query::Arguments::ArgumentValue
+        extract_argument(value, field_defn)
+        extract_arguments(value.value, field_defn)
+      when ::GraphQL::Schema::InputObject
+        extract_arguments(value.arguments.argument_values.values, field_defn)
+      end
```

Comment:
Is there ever any default case? Should we error if it isn't one of these types?

Reply:
Good question! The only scenario that would enter the `else` block, if we had one would be when `value` is an argument scalar value, which we don't care about, since:

```ruby
when ::GraphQL::Query::Arguments::ArgumentValue
  extract_argument(value, field_defn) #  we log all that we need about scalar arguments here
  extract_arguments(value.value, field_defn) # we recurse here anyway, regardless of whether `value.value` is a nested array, hash, set of arguments, argument value or scalar. If it's a scalar, we no-op rather than handling it via `else`.
```